src/analysis/processing/qgsalgorithmsnapgeometries.cpp

```diff
@@ -0,0 +1,411 @@
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  double thresh = parameterAsDouble( parameters, QStringLiteral( "THRESHOLD" ), context );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, source->fields(), source->wkbType(), source->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+
+  // now go and snap vertices of geometries of source together
+  run( *source.get(), *sink.get(), thresh, feedback );
+
+  return outputs;
+}
+
+
+//! record about vertex coordinates and index of anchor to which it is snapped
+typedef struct
+{
+  //! coordinates of the point
+  double x, y;
+
+  /**
+   * Anchor information:
+   *  0+ - index of anchor to which this point should be snapped
+   * -1  - initial value (undefined)
+   * -2  - this point is an anchor, i.e. do not snap this point (snap others to this point)
+   */
+  int anchor;
+} AnchorPoint;
+
+
+//! record about anchor being along a segment
+typedef struct
+{
+  int anchor;    //!< Index of the anchor point
+  double along;  //!< Distance of the anchor point along the segment
+} AnchorAlongSegment;
+
+
+static void buildSnapIndex( QgsFeatureIterator &fi, QgsSpatialIndex &index, QVector<AnchorPoint> &pnts, QgsProcessingFeedback *feedback, int &count, int totalCount )
+{
+  QgsFeature f;
+  int pntId = 0;
+
+  while ( fi.nextFeature( f ) )
+  {
+    QgsGeometry g = f.geometry();
+
+    for ( auto it = g.vertices_begin(); it != g.vertices_end(); ++it )
+    {
+      QgsPoint pt = *it;
+      QgsRectangle rect( pt.x(), pt.y(), pt.x(), pt.y() );
+
+      QList<QgsFeatureId> ids = index.intersects( rect );
+      if ( ids.isEmpty() )
+      {
+        // add to tree and to structure
+        index.insertFeature( pntId, pt.boundingBox() );
+
+        AnchorPoint xp;
+        xp.x = pt.x();
+        xp.y = pt.y();
+        xp.anchor = -1;
+        pnts.append( xp );
+        pntId++;
+      }
+    }
+
+    ++count;
+    feedback->setProgress( 100. * count / totalCount );
+  }
+}
+
+
+static void assignAnchors( QgsSpatialIndex &index, QVector<AnchorPoint> &pnts, double thresh )
+{
+  double thresh2 = thresh * thresh;
+  int nanchors = 0, ntosnap = 0;
+  for ( int point = 0; point < pnts.count(); ++point )
+  {
+    if ( pnts[point].anchor >= 0 )
+      continue;
+
+    pnts[point].anchor = -2; // make it anchor
+    nanchors++;
+
+    // Find points in threshold
+    double x = pnts[point].x, y = pnts[point].y;
+    QgsRectangle rect( x - thresh, y - thresh, x + thresh, y + thresh );
+
+    const QList<QgsFeatureId> ids = index.intersects( rect );
+    for ( QgsFeatureId pointb : ids )
```

Comment:
Maybe `for ( const QgsFeatureId &pointb : ids ){...}`?

Reply:
```const QgsFeatureId &``` is actually worse since QgsFeatureId is a 64-bit int, so it is same length as a reference, but a reference would require extra dereferencing...